File: sdk/python/example/example_sign.py
import base64
import hashlib
import hmac
import json
import logging
import os
import time
import uuid
from urllib.parse import urlencode

import requests


class KcSigner:
    def __init__(self, api_key: str, api_secret: str, api_passphrase: str):
        """
        KcSigner contains information about `apiKey`, `apiSecret`, `apiPassPhrase`
        and provides methods to sign and generate headers for API requests.
        """
        self.api_key = api_key or ""
        self.api_secret = api_secret or ""
        self.api_passphrase = api_passphrase or ""

        if api_passphrase and api_secret:
            self.api_passphrase = self.sign(api_passphrase.encode('utf-8'), api_secret.encode('utf-8'))

        if not all([api_key, api_secret, api_passphrase]):
            logging.warning("API token is empty. Access is restricted to public interfaces only.")

    def sign(self, plain: bytes, key: bytes) -> str:
        hm = hmac.new(key, plain, hashlib.sha256)
        return base64.b64encode(hm.digest()).decode()

    def headers(self, plain: str) -> dict:
        """
        Headers method generates and returns a map of signature headers needed for API authorization
        It takes a plain string as an argument to help form the signature. The outputs are set of API headers.
        """
        timestamp = str(int(time.time() * 1000))
        signature = self.sign((timestamp + plain).encode('utf-8'), self.api_secret.encode('utf-8'))

        return {
            "KC-API-KEY": self.api_key,
            "KC-API-PASSPHRASE": self.api_passphrase,
            "KC-API-TIMESTAMP": timestamp,
            "KC-API-SIGN": signature,
            "KC-API-KEY-VERSION": "2"
        }


def process_headers(signer: KcSigner, body: bytes, raw_url: str, request: requests.PreparedRequest, method: str):
    request.headers["Content-Type"] = "application/json"

    # Create the payload by combining method, raw URL, and body
    payload = method + raw_url + body.decode()
    headers = signer.headers(payload)

    # Add headers to the request
    request.headers.update(headers)


def get_trade_fees(signer: KcSigner, session: requests.Session):
    endpoint = "https://api.kucoin.com"
    path = "/api/v1/trade-fees"
    method = "GET"
    query_params = {"symbols": "BTC-USDT"}

    # Build full URL and raw URL
    full_path = f"{endpoint}{path}?{urlencode(query_params)}"
    raw_url = f"{path}?{urlencode(query_params)}"

    req = requests.Request(method=method, url=full_path).prepare()
    process_headers(signer, b"", raw_url, req, method)

    resp = session.send(req)
    resp_obj = json.loads(resp.content)
    print(resp_obj)


def add_limit_order(signer: KcSigner, session: requests.Session):
    endpoint = "https://api.kucoin.com"
    path = "/api/v1/hf/orders"
    method = "POST"

    # Prepare order data
    order_data = json.dumps({
        "clientOid": str(uuid.uuid4()),
        "side": "buy",
        "symbol": "BTC-USDT",
        "type": "limit",
        "price": "10000",
        "size": "0.001",
    })

    # Build full URL and raw URL
    full_path = f"{endpoint}{path}"
    raw_url = path

    req = requests.Request(method=method, url=full_path, data=order_data).prepare()
    process_headers(signer, order_data.encode(), raw_url, req, method)

    resp = session.send(req)
    resp_obj = json.loads(resp.content)
    print(resp_obj)


if __name__ == '__main__':
    # Load API credentials from environment variables
    key = os.getenv("API_KEY", "")
    secret = os.getenv("API_SECRET", "")
    passphrase = os.getenv("API_PASSPHRASE", "")

    session = requests.Session()
    signer = KcSigner(key, secret, passphrase)

    # Perform operations
    get_trade_fees(signer, session)
    add_limit_order(signer, session)

File: sdk/golang/example/example_sign.go
package main

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"github.com/google/uuid"
	"io"
	"net/http"
	"os"
	"strconv"
	"time"
)

type KcSigner struct {
	apiKey        string
	apiSecret     string
	apiPassPhrase string
}

func Sign(plain []byte, key []byte) []byte {
	hm := hmac.New(sha256.New, key)
	hm.Write(plain)
	data := hm.Sum(nil)
	return []byte(base64.StdEncoding.EncodeToString(data))
}

func (ks *KcSigner) Headers(plain string) map[string]string {
	t := strconv.FormatInt(time.Now().UnixNano()/1000000, 10)
	p := []byte(t + plain)
	s := string(Sign(p, []byte(ks.apiSecret)))
	ksHeaders := map[string]string{
		"KC-API-KEY":         ks.apiKey,
		"KC-API-PASSPHRASE":  ks.apiPassPhrase,
		"KC-API-TIMESTAMP":   t,
		"KC-API-SIGN":        s,
		"KC-API-KEY-VERSION": "2",
	}
	return ksHeaders
}

func NewKcSigner(key, secret, passPhrase string) *KcSigner {
	ks := &KcSigner{
		apiKey:        key,
		apiSecret:     secret,
		apiPassPhrase: string(Sign([]byte(passPhrase), []byte(secret))),
	}
	return ks
}

func getTradeFees(signer *KcSigner, client *http.Client) {
	endpoint := "https://api.kucoin.com"
	path := "/api/v1/trade-fees"
	method := "GET"
	queryParams := "symbols=BTC-USDT"

	fullURL := fmt.Sprintf("%s%s?%s", endpoint, path, queryParams)
	rawPath := fmt.Sprintf("%s?%s", path, queryParams)

	req, err := http.NewRequest(method, fullURL, nil)
	if err != nil {
		fmt.Println("Error creating request:", err)
		return
	}
	var b bytes.Buffer
	b.WriteString(method)
	b.WriteString(rawPath)
	b.Write([]byte{})

	headers := signer.Headers(b.String())
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	resp, err := client.Do(req)
	if err != nil {
		fmt.Println("Error sending request:", err)
		return
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Println("Error reading response:", err)
		return
	}

	fmt.Println("Response:", string(body))
}

func addLimitOrder(signer *KcSigner, client *http.Client) {
	endpoint := "https://api.kucoin.com"
	path := "/api/v1/hf/orders"
	method := "POST"

	orderData := map[string]interface{}{
		"clientOid": uuid.NewString(),
		"side":      "buy",
		"symbol":    "BTC-USDT",
		"type":      "limit",
		"price":     "10000",
		"size":      "0.001",
	}

	orderDataBytes, err := json.Marshal(orderData)
	if err != nil {
		fmt.Println("Error marshalling order data:", err)
		return
	}

	fullURL := fmt.Sprintf("%s%s", endpoint, path)

	var b bytes.Buffer
	b.WriteString(method)
	b.WriteString(path)
	b.Write(orderDataBytes)

	req, err := http.NewRequest(method, fullURL, bytes.NewBuffer(orderDataBytes))
	if err != nil {
		fmt.Println("Error creating request:", err)
		return
	}

	headers := signer.Headers(b.String())
	for k, v := range headers {
		req.Header.Set(k, v)
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		fmt.Println("Error sending request:", err)
		return
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Println("Error reading response:", err)
		return
	}

	fmt.Println("Response:", string(body))
}

func SignExample() {
	apiKey := os.Getenv("API_KEY")
	apiSecret := os.Getenv("API_SECRET")
	passphrase := os.Getenv("API_PASSPHRASE")

	signer := NewKcSigner(apiKey, apiSecret, passphrase)

	client := &http.Client{}

	getTradeFees(signer, client)
	addLimitOrder(signer, client)
}
